doc/man/man1/fields.1
\"    This documentation is copyright 1994 David Huron.
.TH fields 1 "1994 Dec. 4"
.AT 3
.sp 2
.SH "NAME"
.in +2
.in +11
.ti -11
\fBfields\fR  --  list spine/field structure of a Humdrum file
.in -11
.in -2
.sp 1
.sp 1
.SH "SYNOPSIS"
.in +2
.in +7
.ti -7
\fBfields\fR  [\fB-gils\fR]  [\fB-r \fIregexp\fR]  [\fIinputfile ...\fR]  [ \fB> \fIoutputfile.ftf\fR]
.in -7
.in -2
.sp 1
.sp 1
.SH "DESCRIPTION"
.in +2
The
.B "fields"
command works in conjunction with the
.B "-t"
option of the
.B "extract"
command to permit highly selective extraction of data from a Humdrum file.
This command is used only rarely; it is invoked when the more usual options of
.B "extract"
fail to provide sufficient flexibility in isolating certain Humdrum data.
.sp 1
.sp 1
The
.B "fields"
command outputs a so-called \(odfield-trace file\(cd for a given Humdrum input.
This file can be edited by the user and then used with the
.B "extract"
command to select data from the original input file.
For repetitive tasks, the field-trace file is more convenient than
manual editing of the input file.
.sp 1
.sp 1
When invoking the
.B "fields"
command, the user specifies the types of records to be used as reference
points in the editing task.
The command then produces a listing of record numbers as well as the
corresponding number of fields for each record of the specified type.
For example, the user may wish to use spine-path changes as reference
points for editing an input.
.sp 1
.sp 1
Each output record from the
.B "fields"
command consists of three pieces of information, such as illustrated below:
.sp 1
.sp 1
.in +2
13      1-10      # {(4.g#  4.b  ...
.in -2
.sp 1
.sp 1
The first field consists of a single number identifying the corresponding
line number of the input file.
The second field consists of two numbers separated by a dash.
The second number indicates the number of currently active spines.
The number prior to the dash is always 1 (see below).
The third field is a comment beginning with the octothorpe (#) and
continuing with the first ten characters of the corresponding line
from the input file.
The purpose of this comment information is to help orient the user when
editing a field-trace file.
.sp 1
.sp 1
Field-trace information can be requested by record-type:
.B "-g"
for global comments,
.B "-l"
for local comments,
.B "-i"
for exclusive and tandem interpretations, and
.B "-s"
for spine-path indicators.
More than one record-type can be requested.
For example, when the
.B "-gl"
options are invoked, the
.B "fields"
command will produce an output line each time a local or global comment
is encountered in the input.
The line number and number of fields will be given in the output.
.sp 1
.sp 1
For global comments, the number of fields output for the current
line is equal to the number of fields for the most recent non-global
comment record.
For spine-path records, the current record as well as the next
record are output so that the user knows the changes in the number of spines. 
.sp 1
.sp 1
A
.B "-r"
option permits the user to specify a
.I "regular expression;"
field-trace data is output for each record matching the specified
regular expression.
.sp 1
.sp 1
Note that when an unedited field-trace file is used in conjunction with
.B "extract -t,"
the output is identical to the input.
By modifying the field-range (second column in the output), the
user can select which specific fields will be output.
.sp 1
.sp 1
It is recommended that output files produced using the
.B "fields"
command should be given names with the distinguishing `.ftf' extension.
.in -2
.sp 1
.sp 1
.SH "OPTIONS"
.in +2
The
.B "fields"
command provides the following options:
.sp 1
.TS
l l.
\fB-g\fR	identify lines with global comments
\fB-h\fR	displays a help screen summarizing the command syntax
\fB-i\fR	identify any interpretation record
\fB-l\fR	identify lines with local commens
\fB-r \fIregexp\fR	identify all lines matching \fIregexp\fR
\fB-s\fR	identify lines with spine-path indicators
.TE
.sp 1
Options are specified in the command line.
.in -2
.sp 1
.sp 1
.SH "SAMPLE OUTPUTS"
.in +2
Consider by way of illustration the following input file:
.in +2
.sp 1
.TS
l s s l
l l l l.
!! `fields' example
**numbers	**alpha	**symbols
*	*betics	*
14	abc	%@
!some	!local	!comments
3	.	#&
*-	*	*
jkl	$+
.T&
l s s l
l l l l.
!! a global comment
*^	*
uvw	xyz	&%
*	*+	*
*	*	**numbers	*
mno	pqr	87	{^}
*-	*-	*-	*-
.TE
.sp 1
.in -2
With the
.B "-g"
option, the fields command will produce the following output:
.in +2
.sp 1
.TS
l l l.
1	1-1	# Line 1 must appear in the file.
9	1-2	# !! a globa ....
.TE
.sp 1
.in -2
With the
.B "-l"
option, the fields command will produce the following output:
.in +2
.sp 1
.TS
l l l.
1	1-1	# Line 1 must appear in the file.
5	1-3	# !some    !loc ....
.TE
.sp 1
.in -2
With the
.B "-s"
option, the fields command will produce the following output:
.in +2
.sp 1
.TS
l l l.
1	1-1	# Line 1 must appear in the file.
7	1-3	# *-   *    * ....
8	1-2	# Line after path indicator record
10	1-2	# *^   * ....
11	1-3	# Line after path indicator record
12	1-3	# *    *+   * ....
13	1-4	# Line after path indicator record
15	1-4	# *-   *-   *-   * ....
.TE
.sp 1
.in -2
With the
.B "-i"
option, the fields command will produce the following output:
.in +2
.sp 1
.TS
l l l.
1	1-1	# Line 1 must appear in the file.
2	1-3	# **numbers   ....
3	1-3	# *    *betics    ....
7	1-3	# *-   *    * ....
8	1-2	# Line after path indicator record
10	1-2	# *^   * ....
11	1-3	# Line after path indicator record
12	1-3	# *    *+   * ....
13	1-4	# *    *    **numb ....
15	1-4	# *-   *-   *-   * ....
.TE
.sp 1
.in -2
Using the
.B "-r"
option we can specify a regular expression on which record information
cues.
Executing the following command:
.sp 1
.sp 1
.in +2
fields -r '[0-9]' input
.in -2
.sp 1
.sp 1
produces the following output:
.in +2
.sp 1
.TS
l l l.
1	1-1	# Line 1 must appear in the file.
4	1-3	# 14   abc   %@ ....
6	1-3	# 3    .     #& ....
14	1-4	# mno  pqr   87 ....
.TE
.sp 1
.in -2
Notice that only those records containing numerical data tokens
have been listed.
.in -2
.sp 1
.sp 1
.SH "PORTABILITY"
.in +2
\s-1DOS\s+1 2.0 and up, with the \s-1MKS\s+1 Toolkit.
\s-1OS/2\s+1 with the \s-1MKS\s+1 Toolkit.
\s-1UNIX\s+1 systems supporting the
.I "Korn"
shell or
.I "Bourne"
shell command interpreters, and revised
.I "awk"
(1985).
.in -2
.sp 1
.sp 1
.SH "SEE ALSO"
.in +2
\fBextract\fR (1),
\fBregexp\fR (1)
.in -2


toolkit/awk/fields.awk
#####################################################################
##					FIELDS.AWK
# Programmed by: Tim Racinsky         Date: July, 1993
# Copyright (c) 1993 David Huron
#
# Modifications:
#    Date:   Programmer:         	Description:
# May 9/94	Tim Racinsky		Changed -u option to -r
# June 7/94	Tim Racinsky		Modified parse_command for getopts command
#
#
#	This program simply allows the user to obtain a listing of record
# numbers and the corresponding number of fields for that record for
# types of records that the user specifies on the command line.  Each
# line of input is also appended as a comment to each line of output. This
# program is called using 'fields.ksh'.
#	This program will usually be used in conjuntion with the extract2.awk.
#
#	The following gives a brief summary of the available options:
#
#	-g	signify global comments
#	-l	signify local comments
#	-s	sifnify spine-path interpretations only
#	-i	signify all interpretations
#	-r [regexpr] allow user to specify marker
#
#	For global comments, the number of fields output for the current
# line is equal to the number of fields for the most recent non-global
# comment record (held in variable current_no_of_spines).
#	For spine-path records, the current record as well as the next
# record are output so that the user knows the changes in the number of spines. 
#	This program assumes a valid humdrum file as input.
#
# Main Functions used
#	Parse_command()	Store_indiators()		Process_indicators()
#	Determine_print()
#
#	VARIABLES:
#
#  -spine_path_record:  flag to indicate if current interpretation record
#   is also a spine-path record.
#  -USER:  flag to determine if user has specified '-r' option.
#  -REGEXP:  holds regular expression if user specified '-r' option.
#  -options:  holds list of options that user has specified.
#  -print_next:  flag to determine if current record or next non-global
#   comment record should be printed (used for spine-path records).
#  -current_no_of_spines:
#
# (Note that it is possible for one line to be printed more than once!)?????
#
BEGIN {
	USAGE="\nUSAGE: fields -h               (Help Screen)\n       fields "\
		 "[-gils] [-r regexpr] [file ...]"
	TRUE = 1; FALSE = 0
	FS = OFS = "\t"
	#
	# Set standard error to the appropriate 'file'
	#
	#if (ARGV[0] == "nawk") stderr = "/dev/tty"
	#else stderr = "con"
	"echo $CON" | getline stderr
	close("echo $CON")
	USER = FALSE
	REGEXP = ""
	options = ""
	#
	# Determine the options specified by the user
	#
	parse_command()
	}
{
#
# If this is the first record of the current file, print the line
# and reset the variables.
#
if (FNR == 1)
	{
	#
	# The output always includes the first line.
	#
	if ($0 ~ /^!!/)
		{
		print FNR "\t1-1\t# Line 1 must appear in the file."
		current_no_of_spines = 1
		}
	else
		{
		print FNR "\t1-" NF "\t# Line 1 must appear in the file."
		current_no_of_spines = NF
		}
	print_next = FALSE
	spine_path_record = FALSE
	next
	}
#
# If currently looking at a record and it follows a spine-path
# record (and spine-path records are being extracted)
#
if (print_next) determine_print()
#
# This checks for global comments.
#
if ($0 ~ /^!!/)
	{
	if (options ~ /g/)
		{
		print FNR "\t1-" current_no_of_spines "\t# " substr($0,1,10) " ...."
		next
		}
	}
#
# This checks for local comments.
#
else if ($0 ~ /^!/ && options ~ /l/)
	{
	print FNR "\t1-" current_no_of_spines "\t# " substr($0,1,10) " ...."
	next
	}
#
# This checks for interpretation records.
#
if ($0 ~ /^\*/)
	{
	current_no_of_spines = NF
	spine_path_record = FALSE
	for (i = 1; i <= current_no_of_spines; i++)
		{
		if ($i ~ /^\*\+$|^\*\-$|^\*\^$|^\*v$|^\*x$/)
			{
			spine_path_record = TRUE
			break
			}
		}
	#
	# If it is a spine_path_record and the user wants to print it, then
	# print the number of fields of the NEXT line (if it is not a global
	# comment) as well as the current one.
	#
	if (spine_path_record)
		{
		if (options ~ /[is]/)
			{
			print FNR "\t1-" current_no_of_spines "\t# " substr($0,1,10) \
				 " ...."
			print_next = TRUE
			}
		store_indicators()
		process_indicators()
		if (current_no_of_spines == 0)
			{
			current_no_of_spines = 1
			print_next = FALSE
			}
		if (options ~ /[is]/) next
		}
	#
	# Otherwise just print out the record if it was specified.
	#
	else
		{
		if (options ~ /i/)
			{
			print FNR "\t1-" current_no_of_spines "\t# " substr($0,1,10) \
				 " ...."
			next
			}
		}
	}
#
# If the user has specified a 'custom' regular expression, check for it
# and print the line if it exists.
#
if (USER && $0 ~ REGEXP)
	print FNR "\t1-" current_no_of_spines "\t# " substr($0,1,10) " ...."
}

############################################################################
#				FUNCTIONS USED IN FIELDS.AWK
############################################################################

############################################################################
##					Function Parse_command
# This function inspects the array ARGV in order to determine the correct
# options that the user has specified.
#
function parse_command()
	{
	if (ARGV[2] != "null") options = ARGV[2]
	if (ARGV[3] != "null")
		{
		USER = TRUE
		REGEXP = ARGV[3]
		}
	ARGV[1] = ARGV[2] = ARGV[3] = ""
	}

##############################################################
##			Function Store_indicators
#	This function allows the spine-path indicators for the 
# current record to be stored in the array 'path_indicator' so
# that they may be used by process_indicators.
#
function store_indicators(   i)
	{
	for (i = 1; i <= current_no_of_spines; i++)
		{
		if ($i == "*") path_indicator[i] = "*"
		else path_indicator[i] = substr($i,2,1)
		}
	}

##################################################################
##			Function Process_indicators
#	This function simply alters the variable current_no_of_spines
# according to the spine path indicators which are present
#
function process_indicators(    i,temp_spines)
	{
	temp_spines = current_no_of_spines
	i = 1
	while (i <= temp_spines)
		{
		if (path_indicator[i] == "x") ;
		else if (path_indicator[i] == "^") current_no_of_spines += 1
		else if (path_indicator[i] == "+") current_no_of_spines += 1
		else if (path_indicator[i] == "-") current_no_of_spines -= 1
		else if (path_indicator[i] == "v")
			{
			i += 1
			while (path_indicator[i] == "v")
				{
				current_no_of_spines -= 1
				i += 1
				}
			i -= 1
			}
		i += 1
		}
	}

############################################################################
##				Function Determine_print()
#	This function determines whether or not a record following a spine path
# record should be printed or not depending on the current options
#
function determine_print(      i,spine_path_record)
	{
	#
	# Determine if it is another spine path record
	#
	spine_path_record = FALSE
	for (i = 1; i <= NF; i++)
		{
		if ($i ~ /^\*\+$|^\*\-$|^\*\^$|^\*v$|^\*x$/)
			{
			spine_path_record = TRUE
			break
			}
		}
	#
	# If it is global comment and global comments are not being printed
	#
	if ($0 ~ /^!!/)
		{
		if(options !~ /g/)
			{
			print FNR "\t1-" current_no_of_spines "\t# "\
				 "Line after path indicator record"
			}
		}
	#
	# If it is a local comment and local comments are not being printed
	#
	else if ($0 ~ /^![^!]/)
		{
		if (options !~ /l/)
			{
			print FNR "\t1-" current_no_of_spines "\t# "\
				 "Line after path indicator record"
			}
		}
	#
	# If it is an interpretation record
	#
	else if ($0 ~ /^\*/)
		{
		#
		# If it is not a spine path record and interpretations are not
		# being printed
		#
		if (!spine_path_record)
			{
			if (options !~ /i/)
				{
				print FNR "\t1-" current_no_of_spines "\t# "\
					 "Line after path indicator record"
				}
			}
		}
	#
	# If the user is matching regular expressions and this line doesn't
	# match it, print it
	#
	else if (USER)
		{
		if ($0 !~ REGEXP)
			{
			print FNR "\t1-" current_no_of_spines "\t# "\
				 "Line after path indicator record"
			}
		}
	else
		{
		print FNR "\t1-" current_no_of_spines "\t# Line after path "\
			 "indicator record"
		}
	print_next = FALSE
	}
